File: Python/kontikiTests.py
import unittest
import pandas as pd

from kontiki import Kontiki
from publication import Publication


class TestKontiki(unittest.TestCase):
    def test_query_books_valid_general_query_string_succeeds(self):
        kontiki = Kontiki()
        tokens = "bellman richard"
        publications = kontiki.query_books(tokens)
        self.assertTrue(len(publications) > 0)    #  len(publications)

    def test_query_books_invalid_general_query_string_returns_empty_set(self):
        kontiki = Kontiki()
        tokens = "^56m"
        publications = kontiki.query_books(tokens)
        self.assertTrue(len(publications) == 0)  # no publications

    def test_query_books_by_isbn_valid_isbn_succeeds(self):
        kontiki = Kontiki()
        isbn = "9780486432588"
        publication = kontiki.query_books_by_isbn(isbn)
        self.assertIsNotNone(publication)  #  the publication is valid

    def test_query_books_by_isbn_invalid_isbn_returns_empty_set(self):
        kontiki = Kontiki()
        isbn = "9780486432589"
        publication = kontiki.query_books_by_isbn(isbn)
        self.assertIsNone(publication)  # the publication is invalid

    def test_query_books_by_tags_valid_tags_succeeds(self):
        kontiki = Kontiki()
        tags = "mathematics, numerical"
        publications = kontiki.query_books_by_tags(tags)
        self.assertTrue(len(publications) > 0)  # there are publications

    def test_query_books_by_tags_invalid_tags_returns_empty_set(self):
        kontiki = Kontiki()
        tags = "foxilization"
        publications = kontiki.query_books_by_tags(tags)
        self.assertTrue(len(publications) == 0)  # there are no such publications

    def test_retrieve_book_format_bibtex_valid_md5_succeeds(self):
        kontiki = Kontiki()
        md5 = "16D8AEA90AB88BE35C6C8CB7B9335991"
        bibtex = kontiki.retrieve_book(md5)
        self.assertTrue(bibtex is not None)
        self.assertTrue(type(bibtex) is str)

    def test_retrieve_book_format_instance_valid_md5_succeeds(self):
        kontiki = Kontiki()
        md5 = "16D8AEA90AB88BE35C6C8CB7B9335991"
        book = kontiki.retrieve_book(md5, 'instance')
        self.assertTrue(book is not None)
        self.assertTrue(type(book) is Publication)

    def test_retrieve_book_invalid_md5_returns_none(self):
        kontiki = Kontiki()
        md5 = "16D8AEA90AB88BE35C6C8CB7B9335992"
        book = kontiki.retrieve_book(md5)
        self.assertTrue(book is None)

    def test_query_articles_valid_query_string_succeeds(self):
        kontiki = Kontiki()
        tokens = "bellman richard matrix"
        publications = kontiki.query_articles(tokens)
        self.assertTrue(len(publications) > 0)  # df is non-empty

    def test_query_articles_invalid_query_string_returns_empty_set(self):
        kontiki = Kontiki()
        tokens = "bellman richard cats"
        publications = kontiki.query_articles(tokens)
        self.assertTrue(len(publications) == 0)  #  is empty

    def test_retrieve_article_format_bibtex_valid_doi_succeeds(self):
        kontiki = Kontiki()
        doi = "10.1016/0022-247x(78)90262-7"
        bibtex = kontiki.retrieve_article(doi)
        self.assertTrue(bibtex is not None)
        self.assertTrue(type(bibtex) is str)

    def test_retrieve_article_format_instance_valid_doi_succeeds(self):
        kontiki = Kontiki()
        doi = "10.1016/0022-247x(78)90262-7"
        article = kontiki.retrieve_article(doi, 'instance')
        self.assertTrue(article is not None)
        self.assertTrue(type(article) is Publication)

    def test_retrieve_article_invalid_doi_returns_empty_set(self):
        kontiki = Kontiki()
        doi = "10.1016/0022-247x(78)90262-8"
        article = kontiki.retrieve_article(doi)
        self.assertTrue(article is None)

if __name__ == '__main__':
    unittest.main()

File: CSharp/Kontiki.UnitTests/KontikiTests.cs
/***********************************************************************************
* File:         KontikiTests.cs                                                    *
* Contents:     Class KontikiTests                                                 *
* Author:       Stanislav Koncebovski (stanislav@pikkatech.eu)                     *
* Date:         2022-08-08 16:20                                                   *
* Version:      1.0                                                                *
* Copyright:    pikkatech.eu (www.pikkatech.eu)                                    *
***********************************************************************************/
using System;
using System.Collections.Generic;
using NUnit.Framework;

namespace Kontiki.UnitTests
{
	[TestFixture]
	public class KontikiTests
	{
		[Test]
		public void QueryBooksValidGeneralQueryStringSucceeds()
		{
           HapKontikiManager kontiki = PrepareKontikiManager();

			if (kontiki.ConnectionManager.WorkingConnection != null)
			{
				string tokens = "bellman richard";
				List<Publication> publications = kontiki.QueryBooks(tokens);
				Assert.IsNotNull(publications);
				Assert.That(publications.Count > 0);
			}
		}

		[Test]
        public void QueryBooksInvalidGeneralQueryStringReturnsEmptySet()
		{
            HapKontikiManager kontiki = PrepareKontikiManager();

			if (kontiki.ConnectionManager.WorkingConnection != null)
			{
				string tokens		= "^56m";
				List<Publication> publications = kontiki.QueryBooks(tokens);
				Assert.That(publications.Count == 0);
			}
		}

		[Test]
        public void QueryBooksByIsbnValidIsbnSucceeds()
		{
            HapKontikiManager kontiki = PrepareKontikiManager();
			string isbn = "9780486432588";
			List<Publication> books = kontiki.QueryBooks(isbn, QueryColumn.ISBN);
			Assert.That(books.Count == 1);
		}

		[Test]
		public void QueryBooksByIsbnInvalidIsbnReturnsEmptySet()
		{
			HapKontikiManager kontiki = PrepareKontikiManager();
			string isbn = "9780486432580";
			List<Publication> books = kontiki.QueryBooks(isbn, QueryColumn.ISBN);
			Assert.That(books.Count == 0);
		}

		[Test]
		public void QueryBooksByTagsValidTagsSucceeds()
		{
			HapKontikiManager kontiki = PrepareKontikiManager();
			string tags = "mathematics, numerical";
			List<Publication> books = kontiki.QueryBooks(tags, QueryColumn.Tags);
			Assert.That(books.Count > 0);
		}

		[Test]
		public void QueryBooksByTagsInvalidTagsReturnsEmptySet()
		{
			HapKontikiManager kontiki = PrepareKontikiManager();
			string tags = "foxilization";
			List<Publication> books = kontiki.QueryBooks(tags, QueryColumn.Tags);
			Assert.That(books.Count == 0);
		}

		[Test]
		public void RetrieveBookFormatBibtexValidMd5Succeeds()
		{
			HapKontikiManager kontiki = PrepareKontikiManager();
			string md5 = "16D8AEA90AB88BE35C6C8CB7B9335991";
			string bibtex = kontiki.RetrieveBookBibTex(md5);
			Assert.IsNotNull(bibtex);
			Assert.That(bibtex.Length > 0);
		}

		[Test]
		public void RetrieveBookFormatInstanceValidMd5Succeeds()
		{
			HapKontikiManager kontiki = PrepareKontikiManager();
			string md5 = "16D8AEA90AB88BE35C6C8CB7B9335991";
			Publication book = kontiki.RetrieveBook(md5);
			Assert.IsNotNull(book);
		}

		[Test]
		public void RetrieveBookInvalidMd5ReturnsNone()
		{
			HapKontikiManager kontiki = PrepareKontikiManager();
			string md5 = "16D8AEA90AB88BE35C6C8CB7B9335992";
			string bibtex = kontiki.RetrieveBookBibTex(md5);
			Assert.IsNull(bibtex);
		}

		[Test]
		public void QueryArticlesValidQueryStringSucceeds()
		{
			HapKontikiManager kontiki = PrepareKontikiManager();
			string tokens = "bellman richard matrix";
			List<Publication> articles = kontiki.QueryArticles(tokens);
			Assert.That(articles.Count > 0);
		}

		[Test]
		public void QueryArticlesInvalidQueryStringReturnsEmptySet()
		{
			HapKontikiManager kontiki = PrepareKontikiManager();
			string tokens = "bellman richard cats";
			List<Publication> articles = kontiki.QueryArticles(tokens);
			Assert.That(articles.Count == 0);
		}

		[Test]
		public void RetrieveArticleFormatBibtexValidDoiSucceeds()
		{
			HapKontikiManager kontiki = PrepareKontikiManager();
			string doi = "10.1016/0022-247x(78)90262-7";
			Publication article = kontiki.RetrieveArticle(doi);
			Assert.IsNotNull(article);
		}

		[Test]
		public void RetrieveArticleFormatInstanceValidDoiSucceeds()
		{
			HapKontikiManager kontiki = PrepareKontikiManager();
			string doi = "10.1016/0022-247x(78)90262-7";
			string bibtex = kontiki.RetrieveArticleBibTex(doi);
			Assert.IsNotNull(bibtex);
			Assert.That(bibtex.Length > 0);
		}

		[Test]
		public void RetrieveArticleInvalidDoiReturnsEmptySet()
		{
			HapKontikiManager kontiki = PrepareKontikiManager();
			string doi = "10.1016/0022-247x(78)90262-8";
			Publication article = kontiki.RetrieveArticle(doi);
			Assert.IsNull(article);
		}

		#region Private Auxiliary
		private static HapKontikiManager PrepareKontikiManager(string connectionName = "default")
		{
			HapKontikiManager kontiki = new HapKontikiManager();
			string connectionFileName = "kontiki_connections.xml";
			kontiki.ConnectionManager.Load(connectionFileName);
			KontikiConnection connection = kontiki.ConnectionManager.Connections[connectionName];
			kontiki.ConnectionManager.TestConnection(connection);

			return kontiki;
		}
		#endregion
	}
}
